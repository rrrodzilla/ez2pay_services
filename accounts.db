[{"+12063832022":"{\"phone_number\":\"+12063832022\",\"stripe_id\":\"tok_stripe_connect\",\"subscriber\":false}"},{}]