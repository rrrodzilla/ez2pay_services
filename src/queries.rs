#[cynic::query_module(
    schema_path = r#"src/graphql/schema.graphql"#,
    query_module = "query_dsl"
)]
pub mod accounts {

    use crate::query_dsl::*;
    use serde::{Deserialize, Serialize};

    #[derive(cynic::FragmentArguments, Debug)]
    pub struct FindAccountByPhoneArguments {
        pub phone_number: String,
    }

    #[derive(cynic::FragmentArguments, Debug)]
    pub struct FindAccountByIdArguments {
        pub id: cynic::Id,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(
        graphql_type = "Query",
        argument_struct = "FindAccountByPhoneArguments"
    )]
    pub struct FindAccountByPhone {
        #[arguments(phone_number = &args.phone_number)]
        pub find_account_by_phone: Option<Account>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "Query", argument_struct = "FindAccountByIdArguments")]
    pub struct FindAccountById {
        #[arguments(id = args.id.clone())]
        pub find_account_by_id: Option<Account>,
    }

    #[derive(cynic::QueryFragment, Debug, Clone, Serialize, Deserialize)]
    #[cynic(graphql_type = "Account")]
    pub struct Account {
        #[serde(skip, default = "default_id")]
        pub id: cynic::Id,
        pub phone_number: Option<String>,
        pub stripe_id: Option<String>,
    }
    fn default_id() -> cynic::Id {
        cynic::Id::from("")
    }
}
#[cynic::query_module(
    schema_path = r#"src/graphql/schema.graphql"#,
    query_module = "query_dsl"
)]
pub mod products {
    use crate::queries::accounts::*;
    use crate::query_dsl::*;
    use serde::{Deserialize, Serialize};

    #[derive(cynic::FragmentArguments, Debug)]
    pub struct FindProductByIdArguments {
        pub id: cynic::Id,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "Query", argument_struct = "FindProductByIdArguments")]
    pub struct FindProductById {
        #[arguments(id = args.id.clone())]
        pub find_product_by_id: Option<Product>,
    }

    #[derive(cynic::QueryFragment, Debug, Deserialize, Serialize)]
    #[cynic(graphql_type = "Product")]
    pub struct Product {
        pub account: Account,
        pub description: Option<String>,
        pub image: String,
        pub name: Option<String>,
        pub price: i32,
        pub status: ProductStatus,
        pub tax: Option<i32>,
    }

    #[derive(cynic::Enum, Clone, Copy, Debug, Serialize, Deserialize)]
    #[cynic(graphql_type = "ProductStatus")]
    pub enum ProductStatus {
        Published,
        Disabled,
        New,
    }
}
