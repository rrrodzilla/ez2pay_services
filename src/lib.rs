#![allow(dead_code, unused_imports)]
#[macro_use]
extern crate log;
extern crate dotenv;
mod mutations;
mod queries;
mod query_dsl;
mod types;
use crate::mutations::accounts::create::{CreateAccountByPhone, CreateAccountByPhoneArguments};
use crate::queries::accounts::{FindAccountByPhone, FindAccountByPhoneArguments};
use std::env;
use twilio::{Client, OutboundMessage};

pub async fn notify_info(phone_number: &str) {
    //using test values if real values aren't set
    //these vars should all be moved to their own struct
    let sid: &str =
        &env::var("TWILIO_SID").unwrap_or_else(|_| "AC68e25593ac8571dc6b654cec468f67e7".into());
    let secret: &str =
        &env::var("TWILIO_SECRET").unwrap_or_else(|_| "ae037c08815fe4c48d83de8fb71af72b".into());
    let service_phone_number: &str =
        &env::var("EZPAY_PHONE_NUMBER").unwrap_or_else(|_| "+15005550006".into());
    let client = twilio::Client::new(sid, secret);
    match client.send_message(OutboundMessage::new(service_phone_number,phone_number , "To sell a product, text a picture of what you're selling.  Learn more at: https://ez2pay.me")).await {
        Ok(_) => info!("Sent info message to {}", phone_number),
        Err(_) => error!("Couldn't send info message")
    };
}

//we'll want this to return an account id no matter what
//return an existing account or create one from the incoming phone number if it's not found and
//return the new id
pub async fn get_account(phone_number: &str) -> String {
    use cynic::http::SurfExt;
    use cynic::QueryBuilder;

    let db_secret_key: &str =
        &env::var("DB_AUTH_SECRET").unwrap_or_else(|_| panic!("DB_AUTH_SECRET must be set!"));
    let graphql_endpoint: &str = &env::var("GRAPHQL_ENDPOINT")
        .unwrap_or_else(|_| "https://graphql.fauna.com/graphql".into());
    let operation = FindAccountByPhone::build(&FindAccountByPhoneArguments {
        phone_number: phone_number.to_string(),
    });
    //Zm5BRURXRTdoQ0FDQWtRS1cyLXItLTU5MlBnc1hqTDRkTkNfdTJ6VzplenBheTpzZXJ2ZXI=
    let response = surf::post(graphql_endpoint)
        .header("authorization", format!("Basic {}", db_secret_key))
        .header("Accept-Encoding", "gzip, deflate, br")
        .header("Content-Type", "application/json")
        .header("Accept", "application/json")
        .header("Connection", "keep-alive")
        .header("DNT", "1")
        .run_graphql(operation)
        .await
        .unwrap()
        .data
        .unwrap()
        .find_account_by_phone;

    match response {
        Some(a) => {
            let account = a.clone();
            let existing_id = account.id.clone().inner().to_string();
            info!("Retrieved Existing Account: {:?}", existing_id);
            existing_id
        }
        None => {
            //if no account was found, we need to create one
            use cynic::MutationBuilder;
            let operation = CreateAccountByPhone::build(&CreateAccountByPhoneArguments {
                phone_number: phone_number.to_string(),
            });
            let response = surf::post(graphql_endpoint)
                .header("authorization", format!("Basic {}", db_secret_key))
                .header("Accept-Encoding", "gzip, deflate, br")
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .header("Connection", "keep-alive")
                .header("DNT", "1")
                .run_graphql(operation)
                .await
                .unwrap()
                .data
                .unwrap()
                .create_account;
            let new_id = response.id.inner().to_string();
            info!("Created New Account: {:?}", new_id);
            new_id
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn new_account_by_phone_test() {
        use crate::queries::queries::{NewAccountByPhone, NewAccountByPhoneArguments};
        use cynic::MutationBuilder;

        let operation = NewAccountByPhone::build(NewAccountByPhoneArguments {
            phone_number: "+12063832022".to_string(),
        });
        insta::assert_snapshot!(operation.query);
    }
    #[test]
    fn find_account_by_phone_test() {
        use crate::queries::queries::{FindAccountByPhone, FindAccountByPhoneArguments};
        use cynic::QueryBuilder;

        let operation = FindAccountByPhone::build(FindAccountByPhoneArguments {
            phone_number: "+12063832022".to_string(),
        });
        insta::assert_snapshot!(operation.query);
    }
}
