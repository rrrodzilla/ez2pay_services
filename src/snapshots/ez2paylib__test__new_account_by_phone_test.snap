---
source: src/lib.rs
expression: operation.query

---
mutation Mutation($_0: AccountInput!) {
  createAccount(data: $_0) {
    _id
  }
}

